source/fuzz/fuzzer_pass_outline_selection_construct.cpp

```diff
@@ -0,0 +1,152 @@
+    FuzzerContext* fuzzer_context,
+    protobufs::TransformationSequence* transformations)
+    : FuzzerPass(ir_context, transformation_context, fuzzer_context,
+                 transformations) {}
+
+FuzzerPassOutlineSelectionConstruct::~FuzzerPassOutlineSelectionConstruct() =
+    default;
+
+void FuzzerPassOutlineSelectionConstruct::Apply() {
+  for (auto& function : *GetIRContext()->module()) {
+    if (!GetFuzzerContext()->ChoosePercentage(
+            GetFuzzerContext()->GetChanceOfOutliningSelectionConstruct())) {
+      continue;
+    }
+
+    // It is easier to select an element at random from a vector than from an
+    // instruction list.
+    std::vector<opt::BasicBlock*> header_block_candidates;
+    for (auto& block : function) {
+      header_block_candidates.push_back(&block);
+    }
+
+    if (header_block_candidates.empty()) {
+      continue;
+    }
+
+    // Try to get a header block candidate that will increase the chances of the
+    // transformation being applicable.
+    auto* header_block_candidate = MaybeGetHeaderBlockCandidate(
+        header_block_candidates[GetFuzzerContext()->RandomIndex(
+            header_block_candidates)]);
+    if (!header_block_candidate) {
+      continue;
+    }
+
+    std::vector<opt::BasicBlock*> merge_block_candidates;
+    for (auto& block : function) {
+      if (GetIRContext()->GetDominatorAnalysis(&function)->Dominates(
+              header_block_candidate, &block) &&
+          GetIRContext()->GetPostDominatorAnalysis(&function)->Dominates(
+              &block, header_block_candidate)) {
+        merge_block_candidates.push_back(&block);
+      }
+    }
+
+    if (merge_block_candidates.empty()) {
+      continue;
+    }
+
+    // Try to get a merge block candidate that will increase the chances of the
+    // transformation being applicable.
+    auto* merge_block_candidate = MaybeGetMergeBlockCandidate(
+        merge_block_candidates[GetFuzzerContext()->RandomIndex(
+            merge_block_candidates)]);
+    if (!merge_block_candidate) {
+      continue;
+    }
+
+    if (!TransformationOutlineSelectionConstruct::IsApplicableToBlockRange(
+            GetIRContext(), header_block_candidate->id(),
+            merge_block_candidate->id())) {
+      continue;
+    }
+
+    // This boolean constant will be used as a condition for the
+    // OpBranchConditional instruction. We mark it as irrelevant to be able to
+    // replace it with a more interesting value later.
+    FindOrCreateBoolConstant(true, true);
+
+    ApplyTransformation(TransformationOutlineSelectionConstruct(
+        header_block_candidate->id(), merge_block_candidate->id()));
+  }
+}
+
+opt::BasicBlock*
+FuzzerPassOutlineSelectionConstruct::MaybeGetHeaderBlockCandidate(
+    opt::BasicBlock* header_block_candidate) {
+  // Try to create a preheader if |header_block_candidate| is a loop header.
+  if (header_block_candidate->GetLoopMergeInst()) {
```

Comment:
I guess would be better to use `header_block_candidate->IsLoopHeader()` here.

Reply:
Fixed in b755a59.